File: backend/app/schemas/proceso_evaluacion.py
"""
Schemas Pydantic para Gestor de Proceso de Evaluación SEIA (ICSARA/Adendas).
"""
from datetime import date, datetime
from typing import Optional, List, Dict, Any, Literal
from enum import Enum

from pydantic import BaseModel, Field, field_validator


# =============================================================================
# ENUMS
# =============================================================================

class EstadoEvaluacion(str, Enum):
    NO_INGRESADO = "no_ingresado"
    INGRESADO = "ingresado"
    EN_ADMISIBILIDAD = "en_admisibilidad"
    ADMITIDO = "admitido"
    INADMITIDO = "inadmitido"
    EN_EVALUACION = "en_evaluacion"
    ICSARA_EMITIDO = "icsara_emitido"
    ADENDA_EN_REVISION = "adenda_en_revision"
    ICE_EMITIDO = "ice_emitido"
    EN_COMISION = "en_comision"
    RCA_APROBADA = "rca_aprobada"
    RCA_RECHAZADA = "rca_rechazada"
    DESISTIDO = "desistido"
    CADUCADO = "caducado"


class ResultadoRCA(str, Enum):
    FAVORABLE = "favorable"
    FAVORABLE_CON_CONDICIONES = "favorable_con_condiciones"
    DESFAVORABLE = "desfavorable"
    DESISTIMIENTO = "desistimiento"
    CADUCIDAD = "caducidad"


class EstadoICSARA(str, Enum):
    EMITIDO = "emitido"
    RESPONDIDO = "respondido"
    PARCIALMENTE_RESPONDIDO = "parcialmente_respondido"
    VENCIDO = "vencido"


class EstadoAdenda(str, Enum):
    EN_ELABORACION = "en_elaboracion"
    PRESENTADA = "presentada"
    EN_REVISION = "en_revision"
    ACEPTADA = "aceptada"
    CON_OBSERVACIONES = "con_observaciones"
    RECHAZADA = "rechazada"


class ResultadoRevision(str, Enum):
    SUFICIENTE = "suficiente"
    INSUFICIENTE = "insuficiente"
    PARCIALMENTE_SUFICIENTE = "parcialmente_suficiente"


class TipoObservacion(str, Enum):
    AMPLIACION = "ampliacion"
    ACLARACION = "aclaracion"
    RECTIFICACION = "rectificacion"


class PrioridadObservacion(str, Enum):
    CRITICA = "critica"
    IMPORTANTE = "importante"
    MENOR = "menor"


class EstadoObservacionICSARA(str, Enum):
    PENDIENTE = "pendiente"
    RESPONDIDA = "respondida"
    PARCIAL = "parcial"


# =============================================================================
# SCHEMAS DE OBSERVACIONES Y RESPUESTAS
# =============================================================================

class ObservacionICSARA(BaseModel):
    """Observación individual dentro de un ICSARA."""

    id: Optional[str] = Field(None, description="Identificador único, ej: OBS-001 (auto-generado si no se proporciona)")
    oaeca: str = Field(..., description="Organismo que emite la observación, ej: SERNAGEOMIN")
    capitulo_eia: int = Field(..., ge=1, le=11, description="Capítulo del EIA al que refiere")
    componente: Optional[str] = Field(None, description="Componente ambiental específico")
    tipo: TipoObservacion = Field(..., description="Tipo de solicitud")
    texto: str = Field(..., min_length=10, description="Texto de la observación")
    prioridad: PrioridadObservacion = Field(..., description="Nivel de prioridad")
    estado: EstadoObservacionICSARA = Field(
        default=EstadoObservacionICSARA.PENDIENTE,
        description="Estado de la observación"
    )

    model_config = {"use_enum_values": True}


class RespuestaAdenda(BaseModel):
    """Respuesta a una observación dentro de una Adenda."""

    observacion_id: str = Field(..., description="ID de la observación que responde")
    respuesta: str = Field(..., min_length=10, description="Texto de la respuesta")
    anexos_referenciados: List[str] = Field(default_factory=list, description="Códigos de anexos")
    estado: Literal["respondida", "parcial", "no_respondida"] = Field(
        default="respondida",
        description="Estado de la respuesta"
    )
    calificacion_sea: Optional[Literal["suficiente", "insuficiente"]] = Field(
        None,
        description="Calificación del SEA (después de revisión)"
    )

    model_config = {"use_enum_values": True}


class CondicionRCA(BaseModel):
    """Condición de una RCA favorable con condiciones."""

    numero: int = Field(..., description="Número de la condición")
    descripcion: str = Field(..., description="Descripción de la condición")
    plazo: Optional[str] = Field(None, description="Plazo de cumplimiento")
    responsable: Optional[str] = Field(None, description="Responsable del cumplimiento")


# =============================================================================
# SCHEMAS DE CREACIÓN
# =============================================================================

class ProcesoEvaluacionCreate(BaseModel):
    """Datos para iniciar un proceso de evaluación."""

    proyecto_id: int = Field(..., description="ID del proyecto")
    fecha_ingreso: date = Field(..., description="Fecha de ingreso al SEIA")
    plazo_legal_dias: int = Field(default=120, description="Plazo legal (120 EIA, 60 DIA)")

    @field_validator('plazo_legal_dias')
    @classmethod
    def validar_plazo(cls, v):
        if v not in [60, 120]:
            raise ValueError("El plazo debe ser 60 (DIA) o 120 (EIA)")
        return v


class ICSARACreate(BaseModel):
    """Datos para registrar un nuevo ICSARA."""

    numero_icsara: int = Field(..., ge=1, le=3, description="Número de ICSARA (1, 2 o excepcionalmente 3)")
    fecha_emision: date = Field(..., description="Fecha de emisión del ICSARA")
    fecha_limite_respuesta: date = Field(..., description="Fecha límite para presentar Adenda")
    observaciones: List[ObservacionICSARA] = Field(
        default_factory=list,
        description="Lista de observaciones"
    )

    @field_validator('fecha_limite_respuesta')
    @classmethod
    def validar_fecha_limite(cls, v, info):
        fecha_emision = info.data.get('fecha_emision')
        if fecha_emision and v <= fecha_emision:
            raise ValueError("La fecha límite debe ser posterior a la fecha de emisión")
        return v


class AdendaCreate(BaseModel):
    """Datos para registrar una nueva Adenda."""

    numero_adenda: int = Field(..., ge=1, description="Número de Adenda")
    fecha_presentacion: date = Field(..., description="Fecha de presentación de la Adenda")
    respuestas: List[RespuestaAdenda] = Field(
        default_factory=list,
        description="Lista de respuestas a observaciones"
    )


class ActualizarEstadoObservacion(BaseModel):
    """Datos para actualizar el estado de una observación."""

    estado: EstadoObservacionICSARA = Field(..., description="Nuevo estado")
    calificacion: Optional[Literal["suficiente", "insuficiente"]] = Field(
        None,
        description="Calificación del SEA (opcional)"
    )


class RegistrarRCA(BaseModel):
    """Datos para registrar la RCA final."""

    resultado: ResultadoRCA = Field(..., description="Resultado de la RCA")
    numero_rca: str = Field(..., description="Número de la RCA")
    fecha: date = Field(..., description="Fecha de la RCA")
    condiciones: Optional[List[CondicionRCA]] = Field(
        None,
        description="Condiciones (si es favorable con condiciones)"
    )
    url: Optional[str] = Field(None, description="URL del documento RCA")

    model_config = {"use_enum_values": True}


class RegistrarAdmisibilidad(BaseModel):
    """Datos para registrar resultado de admisibilidad."""

    resultado: Literal["admitido", "inadmitido"] = Field(..., description="Resultado")
    fecha: date = Field(..., description="Fecha del pronunciamiento")
    observaciones: Optional[str] = Field(None, description="Observaciones (si inadmitido)")


# =============================================================================
# SCHEMAS DE RESPUESTA
# =============================================================================

class AdendaResponse(BaseModel):
    """Respuesta completa de una Adenda."""

    id: int
    icsara_id: int
    numero_adenda: int
    fecha_presentacion: date
    respuestas: List[Dict[str, Any]]
    total_respuestas: int
    observaciones_resueltas: int
    observaciones_pendientes: int
    porcentaje_resueltas: float
    estado: str
    fecha_revision: Optional[date]
    resultado_revision: Optional[str]
    archivo_id: Optional[int]
    created_at: Optional[datetime]

    model_config = {"from_attributes": True}


class ICSARAResponse(BaseModel):
    """Respuesta completa de un ICSARA."""

    id: int
    proceso_evaluacion_id: int
    numero_icsara: int
    fecha_emision: date
    fecha_limite_respuesta: date
    dias_para_responder: int
    esta_vencido: bool
    observaciones: List[Dict[str, Any]]
    total_observaciones: int
    observaciones_resueltas: int
    observaciones_pendientes: int
    observaciones_por_oaeca: Dict[str, int]
    estado: str
    archivo_id: Optional[int]
    adendas: List[AdendaResponse]
    created_at: Optional[datetime]

    model_config = {"from_attributes": True}


class ProcesoEvaluacionResponse(BaseModel):
    """Respuesta completa de un proceso de evaluación."""

    id: int
    proyecto_id: int
    estado_evaluacion: str
    fecha_ingreso: Optional[date]
    fecha_admisibilidad: Optional[date]
    fecha_rca: Optional[date]
    resultado_rca: Optional[str]
    plazo_legal_dias: int
    dias_transcurridos: int
    dias_suspension: int
    dias_restantes: int
    porcentaje_plazo: float
    numero_rca: Optional[str]
    url_rca: Optional[str]
    condiciones_rca: Optional[List[Dict[str, Any]]]
    total_icsara: int
    total_observaciones: int
    observaciones_pendientes: int
    icsaras: List[ICSARAResponse]
    created_at: Optional[datetime]
    updated_at: Optional[datetime]

    model_config = {"from_attributes": True}


class ResumenProcesoEvaluacion(BaseModel):
    """Resumen del proceso de evaluación para vista rápida."""

    proyecto_id: int
    proyecto_nombre: Optional[str] = None
    estado_actual: str
    estado_label: str
    fecha_ingreso: Optional[date]
    dias_transcurridos: int
    dias_restantes: int
    porcentaje_plazo: float

    # ICSARA
    total_icsara: int
    icsara_actual: Optional[int]
    observaciones_totales: int
    observaciones_pendientes: int
    observaciones_resueltas: int

    # Por OAECA
    observaciones_por_oaeca: Dict[str, int]
    oaeca_criticos: List[str]  # Organismos con más observaciones

    # Adendas
    total_adendas: int
    ultima_adenda_fecha: Optional[date]

    # Próximos pasos
    proxima_accion: str
    fecha_limite: Optional[date]
    alerta: Optional[str]

    model_config = {"from_attributes": True}


class PlazoEvaluacion(BaseModel):
    """Información de plazos del proceso."""

    plazo_legal_dias: int
    dias_transcurridos: int
    dias_suspension: int
    dias_efectivos: int
    dias_restantes: int
    porcentaje_transcurrido: float
    estado_plazo: Literal["normal", "alerta", "critico", "vencido"]
    fecha_limite_estimada: Optional[date]


class EstadoTimeline(BaseModel):
    """Estado para el timeline de evaluación."""

    id: str
    nombre: str
    descripcion: str
    estado: Literal["completado", "actual", "pendiente"]
    fecha: Optional[date]
    icono: Optional[str]


class TimelineEvaluacion(BaseModel):
    """Timeline completo del proceso de evaluación."""

    proyecto_id: int
    estados: List[EstadoTimeline]
    estado_actual: str
    progreso_porcentaje: int


# =============================================================================
# SCHEMAS DE ESTADÍSTICAS
# =============================================================================

class EstadisticasICSARA(BaseModel):
    """Estadísticas de observaciones ICSARA."""

    total_observaciones: int
    por_tipo: Dict[str, int]  # ampliacion, aclaracion, rectificacion
    por_prioridad: Dict[str, int]  # critica, importante, menor
    por_oaeca: Dict[str, int]
    por_capitulo: Dict[int, int]
    por_estado: Dict[str, int]
    oaeca_mas_observaciones: str
    capitulo_mas_observaciones: int


class EstadisticasProceso(BaseModel):
    """Estadísticas generales del proceso."""

    proyecto_id: int
    dias_en_evaluacion: int
    total_icsara: int
    total_adendas: int
    observaciones_totales: int
    observaciones_resueltas: int
    tasa_resolucion: float
    tiempo_promedio_respuesta_dias: Optional[int]

File: frontend/src/types/procesoEvaluacion.ts
/**
 * Tipos para Gestor de Proceso de Evaluacion SEIA (ICSARA/Adendas).
 */

// =============================================================================
// Enums
// =============================================================================

export type EstadoEvaluacion =
  | 'no_ingresado'
  | 'ingresado'
  | 'en_admisibilidad'
  | 'admitido'
  | 'inadmitido'
  | 'en_evaluacion'
  | 'icsara_emitido'
  | 'adenda_en_revision'
  | 'ice_emitido'
  | 'en_comision'
  | 'rca_aprobada'
  | 'rca_rechazada'
  | 'desistido'
  | 'caducado'

export type ResultadoRCA =
  | 'favorable'
  | 'favorable_con_condiciones'
  | 'desfavorable'
  | 'desistimiento'
  | 'caducidad'

export type EstadoICSARA =
  | 'emitido'
  | 'respondido'
  | 'parcialmente_respondido'
  | 'vencido'

export type EstadoAdenda =
  | 'en_elaboracion'
  | 'presentada'
  | 'en_revision'
  | 'aceptada'
  | 'con_observaciones'
  | 'rechazada'

export type ResultadoRevision =
  | 'suficiente'
  | 'insuficiente'
  | 'parcialmente_suficiente'

export type TipoObservacion = 'ampliacion' | 'aclaracion' | 'rectificacion'
export type PrioridadObservacion = 'critica' | 'importante' | 'menor'
export type EstadoObservacionICSARA = 'pendiente' | 'respondida' | 'parcial'
export type EstadoPlazo = 'normal' | 'alerta' | 'critico' | 'vencido'
export type EstadoTimeline = 'completado' | 'actual' | 'pendiente'

// =============================================================================
// Observaciones ICSARA
// =============================================================================

export interface ObservacionICSARA {
  id: string
  oaeca: string
  capitulo_eia: number
  componente?: string | null
  tipo: TipoObservacion
  texto: string
  prioridad: PrioridadObservacion
  estado: EstadoObservacionICSARA
}

export interface ObservacionICSARACreate {
  oaeca: string
  capitulo_eia: number
  componente?: string
  tipo: TipoObservacion
  texto: string
  prioridad: PrioridadObservacion
}

// =============================================================================
// Respuestas Adenda
// =============================================================================

export interface RespuestaAdenda {
  observacion_id: string
  respuesta: string
  anexos_referenciados: string[]
  estado: 'respondida' | 'parcial' | 'no_respondida'
  calificacion_sea?: 'suficiente' | 'insuficiente' | null
}

export interface RespuestaAdendaCreate {
  observacion_id: string
  respuesta: string
  anexos_referenciados?: string[]
  estado?: 'respondida' | 'parcial' | 'no_respondida'
}

// =============================================================================
// Condiciones RCA
// =============================================================================

export interface CondicionRCA {
  numero: number
  descripcion: string
  plazo?: string | null
  responsable?: string | null
}

// =============================================================================
// Adenda
// =============================================================================

export interface Adenda {
  id: number
  icsara_id: number
  numero_adenda: number
  fecha_presentacion: string
  respuestas: RespuestaAdenda[]
  total_respuestas: number
  observaciones_resueltas: number
  observaciones_pendientes: number
  porcentaje_resueltas: number
  estado: EstadoAdenda
  fecha_revision?: string | null
  resultado_revision?: ResultadoRevision | null
  archivo_id?: number | null
  created_at?: string | null
}

export interface AdendaCreate {
  numero_adenda: number
  fecha_presentacion: string
  respuestas?: RespuestaAdendaCreate[]
}

// =============================================================================
// ICSARA
// =============================================================================

export interface ICSARA {
  id: number
  proceso_evaluacion_id: number
  numero_icsara: number
  fecha_emision: string
  fecha_limite_respuesta: string
  dias_para_responder: number
  esta_vencido: boolean
  observaciones: ObservacionICSARA[]
  total_observaciones: number
  observaciones_resueltas: number
  observaciones_pendientes: number
  observaciones_por_oaeca: Record<string, number>
  estado: EstadoICSARA
  archivo_id?: number | null
  adendas: Adenda[]
  created_at?: string | null
}

export interface ICSARACreate {
  numero_icsara: number
  fecha_emision: string
  fecha_limite_respuesta: string
  observaciones?: ObservacionICSARACreate[]
}

// =============================================================================
// Proceso de Evaluacion
// =============================================================================

export interface ProcesoEvaluacion {
  id: number
  proyecto_id: number
  estado_evaluacion: EstadoEvaluacion
  fecha_ingreso?: string | null
  fecha_admisibilidad?: string | null
  fecha_rca?: string | null
  resultado_rca?: ResultadoRCA | null
  plazo_legal_dias: number
  dias_transcurridos: number
  dias_suspension: number
  dias_restantes: number
  porcentaje_plazo: number
  numero_rca?: string | null
  url_rca?: string | null
  condiciones_rca?: CondicionRCA[] | null
  total_icsara: number
  total_observaciones: number
  observaciones_pendientes: number
  icsaras: ICSARA[]
  created_at?: string | null
  updated_at?: string | null
}

// =============================================================================
// Resumen Proceso
// =============================================================================

export interface ResumenProcesoEvaluacion {
  proyecto_id: number
  proyecto_nombre?: string | null
  estado_actual: EstadoEvaluacion
  estado_label: string
  fecha_ingreso?: string | null
  dias_transcurridos: number
  dias_restantes: number
  porcentaje_plazo: number
  total_icsara: number
  icsara_actual?: number | null
  observaciones_totales: number
  observaciones_pendientes: number
  observaciones_resueltas: number
  observaciones_por_oaeca: Record<string, number>
  oaeca_criticos: string[]
  total_adendas: number
  ultima_adenda_fecha?: string | null
  proxima_accion: string
  fecha_limite?: string | null
  alerta?: string | null
}

// =============================================================================
// Plazos
// =============================================================================

export interface PlazoEvaluacion {
  plazo_legal_dias: number
  dias_transcurridos: number
  dias_suspension: number
  dias_efectivos: number
  dias_restantes: number
  porcentaje_transcurrido: number
  estado_plazo: EstadoPlazo
  fecha_limite_estimada?: string | null
}

// =============================================================================
// Timeline
// =============================================================================

export interface EstadoTimelineItem {
  id: string
  nombre: string
  descripcion: string
  estado: EstadoTimeline
  fecha?: string | null
  icono?: string | null
}

export interface TimelineEvaluacion {
  proyecto_id: number
  estados: EstadoTimelineItem[]
  estado_actual: EstadoEvaluacion
  progreso_porcentaje: number
}

// =============================================================================
// Estadisticas
// =============================================================================

export interface EstadisticasICSARA {
  total_observaciones: number
  por_tipo: Record<string, number>
  por_prioridad: Record<string, number>
  por_oaeca: Record<string, number>
  por_capitulo: Record<number, number>
  por_estado: Record<string, number>
  oaeca_mas_observaciones: string
  capitulo_mas_observaciones: number
}

// =============================================================================
// Requests
// =============================================================================

export interface IniciarProcesoRequest {
  fecha_ingreso: string
  instrumento: 'EIA' | 'DIA'
}

export interface RegistrarAdmisibilidadRequest {
  resultado: 'admitido' | 'inadmitido'
  fecha: string
  observaciones?: string
}

export interface RegistrarRCARequest {
  resultado: ResultadoRCA
  numero_rca: string
  fecha: string
  condiciones?: CondicionRCA[]
  url?: string
}

export interface ActualizarEstadoObservacionRequest {
  estado: EstadoObservacionICSARA
  calificacion?: 'suficiente' | 'insuficiente'
}

// =============================================================================
// OAECA (Organismos con Competencia Ambiental)
// =============================================================================

export interface OAECAInfo {
  nombre_completo: string
  competencias: string[]
}

export const OAECAS_MINERIA: Record<string, OAECAInfo> = {
  SERNAGEOMIN: {
    nombre_completo: 'Servicio Nacional de Geologia y Mineria',
    competencias: ['Geologia', 'Relaves', 'Botaderos', 'Plan de cierre minero'],
  },
  DGA: {
    nombre_completo: 'Direccion General de Aguas',
    competencias: ['Recursos hidricos', 'Derechos de agua', 'Hidrogeologia'],
  },
  SAG: {
    nombre_completo: 'Servicio Agricola y Ganadero',
    competencias: ['Flora', 'Fauna', 'Suelos agricolas', 'Subdivision predios'],
  },
  CONAF: {
    nombre_completo: 'Corporacion Nacional Forestal',
    competencias: ['Bosques', 'Areas silvestres protegidas', 'Flora nativa'],
  },
  'SEREMI Salud': {
    nombre_completo: 'Secretaria Regional Ministerial de Salud',
    competencias: ['Residuos', 'Aguas servidas', 'Ruido', 'Calidad del aire'],
  },
  'SEREMI MMA': {
    nombre_completo: 'Secretaria Regional Ministerial del Medio Ambiente',
    competencias: ['Coordinacion', 'Biodiversidad', 'Areas protegidas'],
  },
  CMN: {
    nombre_completo: 'Consejo de Monumentos Nacionales',
    competencias: ['Patrimonio arqueologico', 'Patrimonio historico'],
  },
  CONADI: {
    nombre_completo: 'Corporacion Nacional de Desarrollo Indigena',
    competencias: ['Pueblos indigenas', 'Tierras indigenas', 'CPLI'],
  },
  Municipalidad: {
    nombre_completo: 'Municipalidad respectiva',
    competencias: ['Uso de suelo', 'Planificacion territorial'],
  },
  SERNATUR: {
    nombre_completo: 'Servicio Nacional de Turismo',
    competencias: ['Turismo', 'ZOIT (Zonas de Interes Turistico)'],
  },
}

// =============================================================================
// Helpers
// =============================================================================

export const ESTADO_EVALUACION_LABELS: Record<EstadoEvaluacion, string> = {
  no_ingresado: 'No ingresado',
  ingresado: 'Ingresado',
  en_admisibilidad: 'En admisibilidad',
  admitido: 'Admitido',
  inadmitido: 'Inadmitido',
  en_evaluacion: 'En evaluacion',
  icsara_emitido: 'ICSARA emitido',
  adenda_en_revision: 'Adenda en revision',
  ice_emitido: 'ICE emitido',
  en_comision: 'En comision',
  rca_aprobada: 'RCA aprobada',
  rca_rechazada: 'RCA rechazada',
  desistido: 'Desistido',
  caducado: 'Caducado',
}

export const ESTADO_PLAZO_COLORS: Record<EstadoPlazo, string> = {
  normal: 'success',
  alerta: 'warning',
  critico: 'error',
  vencido: 'error',
}

export const PRIORIDAD_COLORS: Record<PrioridadObservacion, string> = {
  critica: 'error',
  importante: 'warning',
  menor: 'info',
}

export const TIPO_OBSERVACION_LABELS: Record<TipoObservacion, string> = {
  ampliacion: 'Ampliacion',
  aclaracion: 'Aclaracion',
  rectificacion: 'Rectificacion',
}
